soc/espressif/esp32s3/esp32s3-mp.c

```diff
@@ -35,7 +38,135 @@ void esp_appcpu_start(void *entry_point)
+
+	for (int i = 0; i < src_len / 4; i++) {
+		dst[i] = data[i];
+	}
+
+	sys_munmap(data);
+
+	return 0;
+}
+
+int IRAM_ATTR esp_appcpu_image_load(unsigned int hdr_offset, unsigned int *entry_addr)
+{
+	const uint32_t img_off = FIXED_PARTITION_OFFSET(slot0_appcpu_partition);
+	const uint32_t fa_size = FIXED_PARTITION_SIZE(slot0_appcpu_partition);
+	const uint8_t fa_id = FIXED_PARTITION_ID(slot0_appcpu_partition);
+	int rc = 0;
+
+	if (entry_addr == NULL) {
+		ets_printf("Can't return the entry address. Aborting!\n");
+		abort();
+		return -1;
+	}
+
+	ets_printf("Loading appcpu image, area id: %d, offset: 0x%x, hdr.off: 0x%x, size: %d kB\n",
+		   fa_id, img_off, hdr_offset, fa_size / 1024);
+
+	uint32_t mcuboot_header[8] = {0};
+	esp_image_load_header_t image_header = {0};
+
+	const uint32_t *data = (const uint32_t *)sys_mmap(img_off, 0x40);
+
+	memcpy((void *)&mcuboot_header, data, sizeof(mcuboot_header));
+	memcpy((void *)&image_header, data + (hdr_offset / sizeof(uint32_t)),
+	       sizeof(esp_image_load_header_t));
+
+	sys_munmap(data);
+
+	if (image_header.header_magic == ESP_LOAD_HEADER_MAGIC) {
+		ets_printf("MCUboot image format\n");
+	} else if ((image_header.header_magic & 0xff) == 0xE9) {
+		ets_printf("ESP image format is not supported\n");
+		abort();
+	} else {
+		ets_printf("Unknown or empty image detected. Aborting!\n");
+		abort();
+	}
+
+	if (!esp_ptr_in_iram((void *)image_header.iram_dest_addr) ||
+	    !esp_ptr_in_iram((void *)(image_header.iram_dest_addr + image_header.iram_size))) {
+		ets_printf("IRAM region in load header is not valid. Aborting");
+		abort();
+	}
+
+	if (!esp_ptr_in_dram((void *)image_header.dram_dest_addr) ||
+	    !esp_ptr_in_dram((void *)(image_header.dram_dest_addr + image_header.dram_size))) {
+		ets_printf("DRAM region in load header is not valid. Aborting");
+		abort();
+	}
+
+	if (!esp_ptr_in_iram((void *)image_header.entry_addr)) {
+		ets_printf("Application entry point (%xh) is not in IRAM. Aborting",
+			   image_header.entry_addr);
+		abort();
+	}
+
+	ets_printf("IRAM segment: paddr=%08xh, vaddr=%08xh, size=%05xh (%6d) load\n",
+		   (img_off + image_header.iram_flash_offset), image_header.iram_dest_addr,
+		   image_header.iram_size, image_header.iram_size);
+
+	load_segment(img_off + image_header.iram_flash_offset, image_header.iram_size,
+		     image_header.iram_dest_addr);
+
+	ets_printf("DRAM segment: paddr=%08xh, vaddr=%08xh, size=%05xh (%6d) load\n",
+		   (img_off + image_header.dram_flash_offset), image_header.dram_dest_addr,
+		   image_header.dram_size, image_header.dram_size);
+
+	load_segment(img_off + image_header.dram_flash_offset, image_header.dram_size,
+		     image_header.dram_dest_addr);
+
+	ets_printf("Application start=%xh\n", image_header.entry_addr);
+	esp_rom_uart_tx_wait_idle(0);
+
+	assert(entry_addr != NULL);
+	*entry_addr = image_header.entry_addr;
+
+	return rc;
+}
+
+void esp_appcpu_image_stop(void)
+{
+	esp_cpu_stall(1);
```

Comment:
What about replace 1 to ESP32_APPCPU_INDEX or something in that sense?

Reply:
we have 2 main cores 0 and 1, don't think this is necessary.